experimental/integrations/prototype-spark/marquez-spark-agent/src/main/java/marquez/spark/agent/lifecycle/SparkSQLExecutionContext.java

```diff
@@ -1,184 +1,206 @@
-          if (lrr instanceof HadoopFsRelation) {
-            FileIndex location = ((HadoopFsRelation)lrr).location();
-            Collection<Path> rootPaths = JavaConversions.asJavaCollection(location.rootPaths());
-            for (Path rootPath : rootPaths) {
-              result.add(rootPath.toUri().toString());
-            }
-          } else {
-            extraInfo.append("unknown baseRelation in leaf: ").append(leaf.getClass().getName())
-            .append("\n  ").append(leaf)
-            .append("\n  ").append(lrr.getClass().getName()).append(" ").append(lrr).append("\n");
-          }
-        }else {
-          extraInfo.append("unknown leaf node: ").append(leaf.getClass().getName())
-          .append("\n  ").append(leaf).append("\n");
-        }
+  @Override
+  public void end(SparkListenerJobEnd jobEnd) {
+    log.info("Ending job as part of spark-sql:" + jobEnd.jobId());
+    QueryExecution queryExecution = SQLExecution.getQueryExecution(executionId);
+    if (queryExecution == null) {
+      log.info("No execution info {}", queryExecution);
+      return;
+    }
+
+    LineageEvent event =
+        LineageEvent.builder()
+            .inputs(buildInputs(queryExecution.logical()))
+            .outputs(buildOutputs(queryExecution.logical()))
+            .run(
+                buildRun(
+                    buildRunFacets(
+                        buildLogicalPlanFacet(queryExecution.logical()),
+                        buildPhysicalPlanFacet(queryExecution.executedPlan()),
+                        buildJobErrorFacet(jobEnd.jobResult()))))
+            .job(buildJob())
+            .eventTime(toZonedTime(jobEnd.time()))
+            .eventType(getEventType(jobEnd.jobResult()))
+            .producer("org.apache.spark")
+            .build();
+
+    marquezContext.emit(event);
+  }
+
+  protected String getEventType(JobResult jobResult) {
+    if (jobResult.getClass().getSimpleName().startsWith("JobSucceeded")) {
+      return "COMPLETE";
     }
-    return result;
+    return "FAIL";
   }
 
-  private List<String> findOutputs(LogicalPlan logical) {
-    if (logical instanceof InsertIntoHadoopFsRelationCommand) {
-      InsertIntoHadoopFsRelationCommand insert = (InsertIntoHadoopFsRelationCommand)logical;
-      return ImmutableList.of(insert.outputPath().toUri().toString());
-    } else {
-      extraInfo.append("No Output found for plan: ").append(logical.getClass().getName()).append(" ").append(logical).append("\n");
-      return ImmutableList.of();
+  private List<Dataset> buildInputs(LogicalPlan logical) {
+    List<Dataset> inputDatasets = new ArrayList<>();
+    Collection<LogicalPlan> leaves = asJavaCollection(logical.collectLeaves());
+    for (LogicalPlan leaf : leaves) {
+      if (!(leaf instanceof LogicalRelation)) {
+        continue;
+      }
+      LogicalRelation lr = (LogicalRelation) leaf;
+      BaseRelation lrr = lr.relation();
+      if (!(lrr instanceof HadoopFsRelation)) {
+        continue;
+      }
```

Comment:
We should keep track of the leaves we don't understand as they are missing lineage coverage. Possibly this could be added to a facet? for example the type and the content of the leaf node?

Reply:
Change to collect the class name for all leaves. We can expand when we encounter other options.